packages/skeleton/src/components/project.ts

```diff
@@ -0,0 +1,462 @@
+  private services?: RedwoodService[] | undefined
+  private sides?: RedwoodSide[] | undefined
+  private tomls?: RedwoodTOML[] | undefined
+
+  public static getProject({
+    pathWithinProject = '',
+    full = false,
+    readFromCache = true,
+    insertIntoCache = true,
+  }: {
+    pathWithinProject?: string
+    full?: boolean
+    readFromCache?: boolean
+    insertIntoCache?: boolean
+  } = {}) {
+    const projectCache = RedwoodProjectsCache.getInstance().projects
+    const rootPath = getRootPath(pathWithinProject)
+    if (readFromCache) {
+      const cachedProject = projectCache.get(rootPath)
+      if (cachedProject) {
+        return cachedProject
+      }
+    }
+    const project = new RedwoodProject(rootPath, full)
+    if (insertIntoCache) {
+      projectCache.set(project.filepath, project)
+    }
+    return project
+  }
+
+  private constructor(rootPath: string, full = false) {
+    super(rootPath)
+
+    // A project is typescript if we detect a tsconfig.json
+    this.type =
+      fs.existsSync(
+        path.join(getPaths(this.filepath).web.base, 'tsconfig.json')
+      ) ||
+      fs.existsSync(
+        path.join(getPaths(this.filepath).api.base, 'tsconfig.json')
+      )
+        ? RedwoodProjectType.TYPESCRIPT
+        : RedwoodProjectType.JAVASCRIPT
+
+    if (full) {
+      this.cells = extractCells(this)
+      this.directives = extractDirectives(this)
+      this.functions = extractFunctions(this)
+      this.layouts = extractLayouts(this)
+      this.pages = extractPages(this)
+      this.routers = extractRouters(this)
+      this.sdls = extractSDLs(this)
+      this.services = extractServices(this)
+      this.sides = extractSides(this)
+      this.tomls = extractTOMLs(this)
+
+      // Execute the additional checks since we've already extracted all components
+      this.executeAdditionalChecks()
+    }
+  }
+
+  executeAdditionalChecks(): void {
+    this.cells?.forEach((cell) => {
+      cell.executeAdditionalChecks()
+    })
+    this.directives?.forEach((directive) => {
+      directive.executeAdditionalChecks()
+    })
+    this.functions?.forEach((func) => {
+      func.executeAdditionalChecks()
+    })
+    this.layouts?.forEach((layout) => {
+      layout.executeAdditionalChecks()
+    })
+    this.pages?.forEach((page) => {
+      page.executeAdditionalChecks()
+    })
+    this.routers?.forEach((router) => {
+      router.executeAdditionalChecks()
+    })
+    this.sdls?.forEach((sdl) => {
+      sdl.executeAdditionalChecks()
+    })
+    this.services?.forEach((service) => {
+      service.executeAdditionalChecks()
+    })
+    this.sides?.forEach((side) => {
+      side.executeAdditionalChecks()
+    })
+    this.tomls?.forEach((toml) => {
+      toml.executeAdditionalChecks()
+    })
+  }
+
+  getTOMLs(forceExtract = false): RedwoodTOML[] {
+    if (forceExtract || this.tomls === undefined) {
+      this.tomls = extractTOMLs(this)
+    }
+    return this.tomls
+  }
```

Comment:
You may be able to save on duplication here (and in the other `get*` methods) by using deocrators. Though you probably want to wait for TS v5 since it has a different (more accurate to modern JS) decorator implementation https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#decorators